Consolidate go-billy and create new osfs2 (from `fluxcd/pkg/git/gogit/fs`)

pjbgf opened the pull request:
The aim of this PR is to consolidate the go-billy implementation for Flux.

A new FS `osfs2` implementation was introduced, which provides OS filesystem support, without an user-space chroot.
The original (`fluxcd/pkg/git/gogit/fs`) code required some minor changes in order to improve its Windows support, and to align with `osfs` the multi-platform support for operations such as lock/unlock.

pjbgf committed:
osfs2: Create new OS fs implementation
The new osfs2 restricts write operations to descending locations
from the current working dir.

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

pjbgf committed:
build: Upgrade to Go 1.19

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

pjbgf committed:
osfs2: Add JS support
Changes to os_test.go were done due to the conflicts with gomega
dot-import.

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

hiddeco commented:
Think `fluxcd/pkg/fs` refers to `fluxcd/pkg/git/gogit/fs`?

hiddeco commented on README.md:
````diff
@@ -8,14 +8,14 @@ Billy was born as part of [go-git/go-git](https://github.com/go-git/go-git) proj
 ```go
-import "github.com/go-git/go-billy/v5" // with go modules enabled (GO111MODULE=on or outside GOPATH)
-import "github.com/go-git/go-billy" // with go modules disabled
+import "github.com/fluxcd/go-billy/v5" // with go modules enabled (GO111MODULE=on or outside GOPATH)
````
What is the reasoning behind changing the module name? Given our goals are not to create a hard fork, I would say this is a conflicting choice (and signals something different). If the argument is versioning, I think Go Modules `replace` works just fine with tags (`replace github.com/go-git/go-billy/v5 => github.com/fluxcd/go-billy/v5 <tag on fork>`).

pjbgf replied:
It was mostly to align with the go-git approach. I have now reverted this change.

hiddeco replied:
Yes, I was too late with jumping in on that change :grimacing:

aryan9600 commented on osfs2/os.go:
```diff
@@ -0,0 +1,298 @@
+	return os.Readlink(link)
+}
+
+// Chroot returns a new OS filesystem, with working directory set to path.
```
```suggestion
// Chroot returns a new OS filesystem, with the working dir set to the result of joining the provided path with the underlying working dir.
```

pjbgf committed:
osfs2: Improve multi-platform support
The new osfs2 implementation now aligns with osfs for multi-platform
support, including the lock and unlock of files.

The use of "/" was replaced with filepath.Separator. When appropriate,
the tests were updated to use filepath.FromSlash to ensure that paths
support multi-platform testing.

The SecureJoin function was updated to correctly handle volume names
on Windows.

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

pjbgf commented:
Any more changes required for this PR?

hiddeco commented on go.mod:
```diff
@@ -1,10 +1,18 @@
+	gopkg.in/yaml.v3 v3.0.1 // indirect
+)
+
+go 1.19
```
Would put this at top.

pjbgf committed:
Bump dependencies
- github.com/onsi/gomega to version 1.27.2.
- golang.org/x/sys to version 0.5.0.

Signed-off-by: Paulo Gomes <pjbgf@linux.com>

pjbgf commented:
Changes made and bumped dependencies. PTAL

hiddeco reviewed (approved):
Looks good to me, sorry for letting this slip through the cracks :bow:
